src/Results.Immutable/Option.cs

```diff
@@ -32,32 +32,96 @@
     ///     Initializes a new instance of the <see cref="Option{T}" /> struct.
     /// </summary>
     /// <param name="value">Value to wrap as an option.</param>
-    internal Option(T? value)
+    internal Option(T value)
     {
-        Value = value;
-        IsSome = true;
+        some = new Some<T>(value);
     }
 
+    public Some<T>? Some => some;
+
     /// <summary>
     ///     Gets the <see langword="bool" /> indicator,
     ///     which states whether this <see cref="Option{T}" /> wraps an actual value.
     /// </summary>
     /// <seealso cref="IsNone" />
-    public bool IsSome { get; } = false;
+    public bool IsSome => some.HasValue;
 
     /// <summary>
     ///     Gets the <see langword="bool" /> indicator,
     ///     which states whether this <see cref="Option{T}" /> does not have any value.
     /// </summary>
     /// <seealso cref="IsSome" />
-    public bool IsNone => !IsSome;
+    public bool IsNone => !some.HasValue;
+
+    /// <summary>
+    ///     Gets the value associated with this <see cref="Option{T}" />.
+    /// </summary>
+    private readonly Some<T>? some;
+
+    /// <summary>
+    ///     Gets the value associated with this <see cref="Option{T}" />.
+    ///     In case of <see cref="Option{T}.IsNone" />, returns the default value of <typeparamref name="T" />.
+    /// 
+    ///     This property is <strong>unsafe</strong> to use then <typeparamref name="T" /> is a nullable reference type,
+    ///     and non-nullable value types.
+    ///     It is not possible to determine whether the value is null or None.
+    /// </summary>
+    /// <example>
+    ///     This is a safe usage.
+    ///     <code>
+    ///         var option = Option.Some("hello");
+    ///         option.ValueOrDefault; // "hello"
+    ///     </code>
+    /// </example>
+    /// <example>
+    ///     These are not a safe usage.
+    ///     <code>
+    ///         var option = Option.Some&lt;string&gt;(null);
+    ///         option.ValueOrDefault; // null, same as Option.None&lt;string&gt;()
+    ///         var option2 = Option.None&lt;int&gt;();
+    ///         option2.ValueOrDefault; // 0
+
+    ///     </code>
+    /// </example>
+    /// <value>The value associated with this <see cref="Option{T}" /> or default value of <typeparamref name="T" />.</value>
+    public T? ValueOrDefault => some is { Value: var value } ? value : default;
```

Comment:
Slightly more obscure and concise pattern-match:
```suggestion
    public T? ValueOrDefault => some is var (value) ? value : default;
```

Can be re-used throughout the struct's implementation.